units/src/amount.rs

```diff
@@ -186,18 +223,86 @@ impl std::error::Error for ParseAmountError {
     fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
         use ParseAmountError::*;
 
-        match *self {
-            Negative | TooBig | TooPrecise | InvalidFormat | InputTooLarge
+        match self {
+            TooPrecise | InvalidFormat | InputTooLarge
             | InvalidCharacter(_) => None,
-            InvalidDenomination(ref e) => Some(e),
+            OutOfRange(error) => Some(error),
+        }
+    }
+}
+
+/// Returned when a parsed amount is too big or too small.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct OutOfRangeError {
+    is_signed: bool,
+    is_greater_than_max: bool,
+}
+
+impl OutOfRangeError {
+    /// Returns the minimum and maximum allowed values for the type that was parsed.
+    ///
+    /// This can be used to give a hint to the user which values are allowed.
+    pub fn valid_range(&self) -> (i64, u64) {
+        match self.is_signed {
+            true => (i64::MIN, i64::MAX as u64),
+            false => (0, u64::MAX),
+        }
+    }
+
+    /// Returns true if the input value was large than the maximum allowed value.
+    pub fn is_above_max(&self) -> bool {
+        self.is_greater_than_max
+    }
+
+    /// Returns true if the input value was smaller than the minimum allowed value.
+    pub fn is_below_min(&self) -> bool {
+        !self.is_greater_than_max
+    }
+
+    pub(crate) fn too_big(is_signed: bool) -> Self {
+        Self {
+            is_signed,
+            is_greater_than_max: true,
+        }
+    }
+
+    pub(crate) fn too_small() -> Self {
+        Self {
+            // implied - negative() is used for the other
+            is_signed: true,
+            is_greater_than_max: false,
+        }
+    }
+
+    pub(crate) fn negative() -> Self {
+        Self {
+            // implied - too_small() is used for the other
+            is_signed: false,
+            is_greater_than_max: false,
         }
     }
 }
 
-impl From<ParseDenominationError> for ParseAmountError {
-    fn from(e: ParseDenominationError) -> Self { Self::InvalidDenomination(e) }
+impl fmt::Display for OutOfRangeError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.is_greater_than_max {
+            write!(f, "the amount is greater than {}", self.valid_range().1)
+        } else {
+            write!(f, "the amount is less than {}", self.valid_range().0)
+        }
+    }
 }
 
+#[cfg(feature = "std")]
+impl std::error::Error for OutOfRangeError {}
```

Comment:
@Kixunil I put literally hours into repeatedly opening every file in the codebase that contains error code to get all the error code exactly the same so that it becomes boilerplate and devs don't have to look at it. This `{}` is different (in style not functionality). Can we please have a uniform policy? I don't care which it is but if you want to do this can you change every instance of struct errors (or create an issue that describes that and I'll do it).

Reply:
Also we can then reasonably expect newer devs to use the correct style because its uniform.

Reply:
I honestly don't believe it matters and for no-item traits I personally prefer using `{}` because it's shorter and no less clear than whatever else we could have.

Reply:
Ok, I'm going to mentally move to using `{}` for structs, but keep explicit match for enums. Codebase wide will be an all red diff that won't merge conflict too much so I'll do that at some stage when I'm too tired to do anything else.